dev/io.openliberty.security.common.jwt/src/io/openliberty/security/common/jwt/jws/JwsVerificationKeyHelper.java

```diff
@@ -0,0 +1,152 @@
+import java.io.UnsupportedEncodingException;
+import java.security.Key;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.apache.http.NameValuePair;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.HttpGet;
+import org.jose4j.jws.JsonWebSignature;
+import org.jose4j.jwx.JsonWebStructure;
+import org.jose4j.keys.HmacKey;
+
+import com.ibm.websphere.ras.ProtectedString;
+import com.ibm.websphere.ras.annotation.Sensitive;
+import com.ibm.ws.security.common.http.HttpUtils;
+import com.ibm.ws.security.common.jwk.impl.JWKSet;
+import com.ibm.ws.security.common.jwk.impl.JwKRetriever;
+
+import io.openliberty.security.common.jwt.exceptions.SharedKeyMissingException;
+import io.openliberty.security.common.jwt.exceptions.UnsupportedSignatureAlgorithmException;
+import io.openliberty.security.common.jwt.jwk.RemoteJwkData;
+
+public class JwsVerificationKeyHelper {
+
+    public static final Set<String> SUPPORTED_SIGNATURE_ALGORITHMS;
+
+    static {
+        Set<String> supportedAlgorithms = new HashSet<>();
+        supportedAlgorithms.add("RS256");
+        supportedAlgorithms.add("RS384");
+        supportedAlgorithms.add("RS512");
+        supportedAlgorithms.add("HS256");
+        supportedAlgorithms.add("HS384");
+        supportedAlgorithms.add("HS512");
+        supportedAlgorithms.add("ES256");
+        supportedAlgorithms.add("ES384");
+        supportedAlgorithms.add("ES512");
+        SUPPORTED_SIGNATURE_ALGORITHMS = Collections.unmodifiableSet(supportedAlgorithms);
+    }
+
+    private String clientId;
+    @Sensitive
+    private ProtectedString clientSecret;
+    private RemoteJwkData remoteJwkData;
+    private JWKSet jwkSet;
+
+    private JwsVerificationKeyHelper(Builder builder) {
+        this.clientId = builder.clientId;
+        this.clientSecret = builder.clientSecret;
+        this.remoteJwkData = builder.remoteJwkData;
+        this.jwkSet = builder.jwkSet;
+    }
+
+    public Key getVerificationKey(JsonWebStructure jws) throws Exception {
+        String signatureAlgorithmFromJws = getSignatureAlgorithmFromJws(jws);
+        if (signatureAlgorithmFromJws != null && signatureAlgorithmFromJws.startsWith("HS")) {
+            return getSharedKey();
+        }
+        return retrievePublicKey(jws, signatureAlgorithmFromJws);
+    }
+
+    String getSignatureAlgorithmFromJws(JsonWebStructure jws) throws UnsupportedSignatureAlgorithmException {
+        String algorithm = null;
+        if (jws != null && jws instanceof JsonWebSignature) {
+            algorithm = ((JsonWebSignature) jws).getAlgorithmHeaderValue();
+        }
+        if (SUPPORTED_SIGNATURE_ALGORITHMS.contains(algorithm)) {
+            return algorithm;
+        }
+        throw new UnsupportedSignatureAlgorithmException(algorithm);
+    }
+
+    Key getSharedKey() throws SharedKeyMissingException, UnsupportedEncodingException {
```

Comment:
Please confirm the Key value is not traced here or elsewhere since this is the client secret.

Reply:
Confirmed that this only traces the object's address:
```
[11/1/22, 9:24:17:270 CDT] 00000059 id=e443f4a4 openliberty.security.common.jwt.jws.JwsVerificationKeyHelper < getSharedKey Exit  
                                                                                                               org.jose4j.keys.HmacKey@343bb1
```
The client secret is not being traced anywhere with the logging settings we currently have in the FATs.